tests/unit/lms/views/test_application_instance.py

```diff
@@ -1,25 +1,68 @@
+import pytest
+
 from lms.models import ApplicationInstance
-from lms.views.application_instances import create_application_instance
+from lms.views.application_instances import (
+    create_application_instance,
+    new_application_instance,
+)
 
 
-class TestApplicationInstance:
+class TestCreateApplicationInstance:
     def test_it_creates_an_application_instance(self, pyramid_request):
+        create_application_instance(pyramid_request)
+
+        ai = pyramid_request.db.query(ApplicationInstance).one()
+        assert ai.lms_url == "canvas.example.com"
+        assert ai.requesters_email == "email@example.com"
+        assert ai.developer_key is None
+        assert ai.developer_secret is None
+
+    def test_it_saves_the_Canvas_developer_key_and_secret_if_given(
+        self, pyramid_request
+    ):
+        pyramid_request.params["developer_key"] = "example_key"
+        pyramid_request.params["developer_secret"] = "example_secret"
+
+        create_application_instance(pyramid_request)
+
+        ai = pyramid_request.db.query(ApplicationInstance).one()
+        assert ai.developer_key == "example_key"
+        assert ai.developer_secret
+
+    def test_if_developer_key_is_missing_it_doesnt_save_developer_secret_either(
+        self, pyramid_request
+    ):
+        pyramid_request.params["developer_secret"] = "example_secret"
+
+        create_application_instance(pyramid_request)
+
+        ai = pyramid_request.db.query(ApplicationInstance).one()
+        assert ai.developer_key is None
+        assert ai.developer_secret is None
+
+    def test_if_developer_secret_is_missing_it_doesnt_save_developer_key_either(
+        self, pyramid_request
+    ):
+        pyramid_request.params["developer_key"] = "example_key"
+
+        create_application_instance(pyramid_request)
+
+        ai = pyramid_request.db.query(ApplicationInstance).one()
+        assert ai.developer_key is None
+        assert ai.developer_secret is None
```

Comment:
These two tests look like they could be a parameterised `test_if_developer_key_or_secret_is_missing_the_other_isnt_saved` or something by iterating over the key to change.

Reply:
Done